kotlinx-coroutines-core/common/src/flow/operators/Errors.kt

```diff
@@ -67,13 +71,19 @@ public fun <T> Flow<T>.retry(
                 }
                 break
             } catch (e: Throwable) {
-                if (fromDownstream) throw e
+                if (fromDownstream || e.isCancellationCause()) throw e
                 if (!predicate(e) || retries-- == 0) throw e
             }
         }
     }
 }
 
+private suspend fun Throwable.isCancellationCause(): Boolean {
+    val job = coroutineContext[Job]
```

Comment:
Make `coroutineContext` a parameter to this function and drop `suspend` modifier from it to save suspension point in callers of this function.